src/peft/tuners/tuners_utils.py

```diff
@@ -0,0 +1,240 @@
+# Copyright 2023-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from abc import ABC, abstractmethod
+from typing import Any
+
+from torch import nn
+
+from ..config import PeftConfig
+from ..utils import _get_submodules
+
+
+logger = logging.getLogger(__name__)
+
+
+class BaseTuner(nn.Module, ABC):
+    r"""
+    A base tuner model that provides the common methods and attributes for all tuners that are injectable into a
+    torch.nn.Module
+
+    For adding a new Tuner class, one needs to overwrite the following methods:
+
+    - **_prepare_adapter_config**:
+        A private method to eventually prepare the adapter config, for example in case the field `target_modules` is
+        missing.
+    - **_check_target_module_exists**:
+        A helper private method to check if the passed module's key name matches any of the target modules in the
+        adatper_config.
+    - **_create_and_replace**:
+        A private method to create and replace the target module with the adapter module.
+    - **_check_target_module_exists**:
+        A private helper method to check if the passed module's key name matches any of the target modules in the
+        adatper_config.
+
+    The easiest is to check what is done in the `peft.tuners.lora.LoraModel` class.
+
+    Attributes:
+        model (`torch.nn.Module`):
+            The model to which the adapter tuner layers will be attached.
+        forward (`Callable`):
+            The forward method of the model.
+        peft_config (`dict[str, PeftConfig]`):
+            The adapter configuration object, it should be a dictionary of `str` to `PeftConfig` objects.
+        config (`dict[str, Any]`):
+            The model configuration object, it should be a dictionary of `str` to `Any` objects.
+    """
+
+    def __init__(self, model, peft_config, adapter_name):
+        super().__init__()
+
+        self.model = model
+        self.forward = self.model.forward
+
+        # For advanced developpers, if you want to attach multiple adapters to your
+        # model, just add a `peft_config` attribute to your model.
+        if not hasattr(self, "peft_config"):
+            self.peft_config = peft_config
+        else:
+            logger.info(
+                "Already found a `peft_config` attribute in the model. This will lead to having multiple adapters"
+                " in the model. Make sure to know what you are doing!"
+            )
```

Comment:
post this you have to add the new config in the `peft_config` dict, else `inject_adapter` will lead to key not found error

Reply:
Addressed it in  8053c88